File: backend/app/schemas/sale_return.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.customer import CustomerResponse
from app.schemas.customer_collection import UserNestedResponse
from app.schemas.product import ProductResponse


class SaleNestedResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    invoice_no: str
    grand_total: float
    paid_amount: float
    due_amount: float


class SaleReturnItemCreate(BaseModel):
    product_id: str = Field(..., description="Target Product UUID")
    quantity: float = Field(..., gt=0.0, description="Returned Quantity (> 0)")
    unit_price: float = Field(..., ge=0.0, description="Unit Price ($)")

    @field_validator("quantity")
    def validate_quantity(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("Returned quantity must be greater than zero")
        return v


class SaleReturnItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sale_return_id: str
    product_id: str
    quantity: float
    unit_price: float
    total_price: float
    created_at: datetime
    updated_at: datetime

    product: Optional[ProductResponse] = None


class SaleReturnCreate(BaseModel):
    sale_id: Optional[str] = Field(None, description="Linked Sale Invoice UUID")
    customer_id: str = Field(..., description="Target Customer UUID")
    return_date: Optional[datetime] = Field(None, description="Return Transaction Date")
    refund_amount: float = Field(0.0, ge=0.0, description="Cash refund amount paid back to customer ($)")
    reason: Optional[str] = Field(None, description="Return reason/notes")
    items: List[SaleReturnItemCreate] = Field(..., min_length=1, description="List of returned line items")

    @field_validator("items")
    def validate_items(cls, v: List[SaleReturnItemCreate]) -> List[SaleReturnItemCreate]:
        if not v or len(v) == 0:
            raise ValueError("Sale return voucher must contain at least one line item")
        return v


class SaleReturnUpdate(BaseModel):
    refund_amount: Optional[float] = Field(None, ge=0.0)
    reason: Optional[str] = None
    items: Optional[List[SaleReturnItemCreate]] = None


class SaleReturnResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    return_no: str
    sale_id: Optional[str] = None
    customer_id: str
    user_id: str
    return_date: datetime
    grand_total: float
    refund_amount: float
    reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    customer: Optional[CustomerResponse] = None
    sale: Optional[SaleNestedResponse] = None
    user: Optional[UserNestedResponse] = None
    items: List[SaleReturnItemResponse] = []


class SaleReturnableItem(BaseModel):
    product_id: str
    product_name: str
    product_code: str
    unit: str
    sold_quantity: float
    previously_returned_qty: float
    returnable_qty: float
    unit_price: float


class SaleReturnableSummary(BaseModel):
    sale_id: str
    invoice_no: str
    sale_date: datetime
    customer_id: str
    customer_name: str
    customer_phone: str
    customer_code: str
    grand_total: float
    paid_amount: float
    due_amount: float
    items: List[SaleReturnableItem]


class DailyReturnBreakdown(BaseModel):
    date: str
    total_amount: float
    count: int


class SaleReturnReportSummaryData(BaseModel):
    total_returns_count: int
    total_returned_amount: float
    total_refund_amount: float
    today_amount: float
    yesterday_amount: float
    this_week_amount: float
    this_month_amount: float
    daily_breakdown: List[DailyReturnBreakdown]

File: frontend/src/types/sale_return.ts
import { CustomerItem } from "./customer";
import { ProductItem } from "./product";
import { UserItem } from "./user";
import { SaleItem } from "./sale";

export interface SaleReturnItemModel {
  id: string;
  sale_return_id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  total_price: number;
  created_at: string;
  updated_at: string;

  product?: ProductItem;
}

export interface SaleReturnItem {
  id: string;
  return_no: string;
  sale_id?: string | null;
  customer_id: string;
  user_id: string;
  return_date: string;
  grand_total: number;
  refund_amount: number;
  reason?: string | null;
  created_at: string;
  updated_at: string;

  customer?: CustomerItem;
  sale?: Partial<SaleItem>;
  user?: Partial<UserItem>;
  items: SaleReturnItemModel[];
}

export interface SaleReturnItemCreatePayload {
  product_id: string;
  quantity: number;
  unit_price: number;
}

export interface SaleReturnCreatePayload {
  sale_id?: string | null;
  customer_id: string;
  return_date?: string | null;
  refund_amount?: number;
  reason?: string | null;
  items: SaleReturnItemCreatePayload[];
}

export interface SaleReturnUpdatePayload {
  refund_amount?: number;
  reason?: string | null;
  items?: SaleReturnItemCreatePayload[];
}

export interface SaleReturnableItem {
  product_id: string;
  product_name: string;
  product_code: string;
  unit: string;
  sold_quantity: number;
  previously_returned_qty: number;
  returnable_qty: number;
  unit_price: number;
}

export interface SaleReturnableSummary {
  sale_id: string;
  invoice_no: string;
  sale_date: string;
  customer_id: string;
  customer_name: string;
  customer_phone: string;
  customer_code: string;
  grand_total: number;
  paid_amount: number;
  due_amount: number;
  items: SaleReturnableItem[];
}

export interface DailyReturnBreakdown {
  date: string;
  total_amount: number;
  count: number;
}

export interface SaleReturnReportSummaryData {
  total_returns_count: number;
  total_returned_amount: number;
  total_refund_amount: number;
  today_amount: number;
  yesterday_amount: number;
  this_week_amount: number;
  this_month_amount: number;
  daily_breakdown: DailyReturnBreakdown[];
}
